src/hal/motor.cpp

```diff
@@ -2,20 +2,461 @@
+#define MOTOR_STATE_HOLD  1
+#define MOTOR_STATE_OPEN  2
+#define MOTOR_STATE_CLOSE 3
+
+#define MOTOR_DIRECTION_OPEN   -1
+#define MOTOR_DIRECTION_CLOSE   1
+
+// TODO
+// This value should be derived from Timer 3 and Timer 4 frequencies.
+#define MOTOR_CONTROL_STEP_ERROR 5
+
+//**************************************
+// Motor Control Variables
+//**************************************
+
+// Staging area for motor move commands.
+static struct {
+  bool    valid = false;
+  uint8_t direction = 0;
+  uint8_t distance = 0;
+} motor_move_command;
+
+static volatile struct {
+  int32_t step_command = 0;
+  int32_t step_position = 0;
+  int8_t  microstep_position = 0;
+} motor_control;
+
+// Motor State Machine
+static volatile uint8_t motor_state = MOTOR_STATE_OFF;
+
+static volatile bool motor_busy = false; 
+
+//*****************************************************************************
+// Timer 3 Configuration: Motor State Control
+//
+// Configure Timer 3 to generate an interrupt to control motor state.
+// WGM3[3:0] = 0b1001 (PWM, Phase and Frequency Correct Mode, TOP: OCR3A)
+// CS3[2:0] = 0b010 (clk/8 prescaler)
+//*****************************************************************************
+void timer3_setup() {
+  TCCR3A = (0<<WGM31) | (1<<WGM30);
+  TCCR3B = (1<<WGM33) | (0<<WGM32);
+  TCNT3 = 0;
+  // OCR3A = 20000; // 50 Hz (1,000,000 / 20,000 = 50)
+  OCR3A = 10000; // 100 Hz (1,000,000 / 10,000 = 100)
+  TIMSK3 = (1<<OCIE3A);
+  TCCR3B |= (0<<CS32)   | (1<<CS31)  | (0<<CS30);
+}
+
+//*****************************************************************************
+// Timer 4 Configuration: Motor Movement and Position Tracking
+//
+// Configure Timer 4 to generate a PWM output to the motor to cause it to move
+// one (micro)step per pulse. The PWM output is enabled and disabled
+// dynamically, so there are separate *_enabled and *_disabled values for
+// TCCR4B. Configure an interrupt to occur for every PWM pulse to track the
+// motor position.
+//
+// WGM4[3:0] = 0b1001 (PWM, Phase and Frequency Correct Mode, TOP: OCR4A)
+// COM4B[1:0] = 0b10 (Enable PWM output on OC4B -> ATMega 2560 PH4 -> Arduino D7)
+// CS3[2:0] = 0b010 (clk/8 prescaler)
+//
+// Create values for the control registers such that the timer can be enabled
+// and disabled dynamically.
+//*****************************************************************************
+static uint8_t TCCR4A_value = (1<<COM4B1) | (0<<COM4B0) | (0<<WGM41) | (1<<WGM40);
+static uint8_t TCCR4B_value_disabled = (1<<WGM43)  | (0<<WGM42);
+static uint8_t TCCR4B_value_enabled = TCCR4B_value_disabled | (0<<CS42)   | (1<<CS41)  | (0<<CS40);
+
+void timer4_setup()
+{
+  TCCR4A = TCCR4A_value;
+  TCCR4B = TCCR4B_value_disabled;
+  TCNT4 = 0;
+  // OCR4A = 1000; // 1 KHz (1,000,000 / 1,000 = 1,000)
+  // OCR4A = 2000; // 500 Hz (1,000,000 / 2,000 = 500)
+  // OCR4A = 2500; // 400 Hz (1,000,000 / 2,500 = 400)
+  
+  OCR4A = 625;
+  // OCR4A = 375;
+  // OCR4A = 225;
+  // OCR4A = 125;
```

Comment:
Doe these three comments at the top really need to be here? I think it might be better to put the frequencies that it would actually attain below like this:
```suggestion
  
  // OCR4A = 375;  // ~2.66 KHz (1,000,000 / 375 = 2666.7)
  // OCR4A = 225;  // ~4.44 KHz (1,000,000 / 225 = 4444.5)
  // OCR4A = 125;  // 8 KHz (1,000,000 / 125 = 8000)
  OCR4A = 625;  // 1.6 KHz (1,000,000 / 625 = 1600)
```

Reply:
Comments removed.